### PR TITLE
Blended L1+L2 Surface Loss: MSE auxiliary for large-error penalty

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The baseline uses pure L1 (MAE) for the surface loss, which matches the evaluation metric. However, L1 has a constant gradient magnitude regardless of error size — a node with 10x error gets the same gradient magnitude as one with 1x error. Adding a small L2 (MSE) component creates a **quadratic penalty on large errors**, directing more gradient signal to the worst-predicted nodes (suction peaks, stagnation points, TE pressure recovery).

This is different from Huber loss (PR #2236, failed) which REPLACED L1 with L2 for large errors, weakening the penalty. Here we ADD L2 on top of L1, strengthening it.

**Expected mechanism:**
- L1 preserves metric alignment (MAE is what we optimize for)
- L2 component at alpha=0.1 adds 10% MSE penalty → nodes with 3x error get 30% more total gradient vs pure L1
- This should reduce worst-case predictions and improve MAE through variance reduction

### 1. Add CLI flags to `Config` dataclass (near other loss flags):
```python
l2_surface_weight: float = 0.0       # weight for auxiliary L2 (MSE) surface loss; 0=pure L1
```

### 2. Modify the surface loss computation
Find where the surface MAE loss is computed. After the existing L1 loss calculation, add the L2 component:

```python
# Existing L1 surface loss (keep as-is)
surface_l1_loss = ...  # the current MAE computation

# Add L2 component if enabled
if cfg.l2_surface_weight > 0:
    surface_l2_loss = ((pred_surface - gt_surface) ** 2 * surface_mask).sum() / surface_mask.sum().clamp(min=1)
    surface_loss = surface_l1_loss + cfg.l2_surface_weight * surface_l2_loss
else:
    surface_loss = surface_l1_loss
```

**Important details:**
- Apply L2 to the SAME surface nodes as L1 (same mask, same normalization)
- Apply to BOTH fore-foil and aft-foil surface losses
- Do NOT apply L2 to the DCT frequency loss or volume loss
- The L2 should operate on the same normalized targets as L1 (after per-sample-std normalization)
- Log `surface_l2_loss` to W&B separately so we can monitor it

### 3. Run configuration
```bash
cd cfd_tandemfoil && python train.py --agent alphonse --wandb_name "alphonse/blended-l1l2-alpha01" \
  --wandb_group "blended-l1l2" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --l2_surface_weight 0.1 \
  --seed 42
```
Run with seeds 42 and 73.

## Baseline (PR #2213, 2-seed avg)
| Metric | Value | Target to beat |
|--------|-------|---------------|
| p_in | 11.979 | < 11.98 |
| p_oodc | 7.643 | < 7.65 |
| p_tan | 28.341 | < 28.34 |
| p_re | 6.300 | < 6.30 |

Baseline W&B runs: hgml7i2r (s42), qic03vrg (s73)

Reproduce baseline:
```bash
cd cfd_tandemfoil && python train.py --agent alphonse --wandb_name "alphonse/baseline-wake-deficit" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Results

### Blended L1+L2 Surface Loss (alpha=0.1)

**Configuration:** Added `l2_surface_weight=0.1` — 10% MSE penalty on top of the existing L1 surface loss, applied to all 3 channels (Ux, Uy, p) at surface nodes for both fore-foil and aft-foil, after the aft-foil refinement head.

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Target | Verdict |
|--------|---------|---------|------------|----------|--------|---------|
| p_in | 12.3 | 11.6 | **11.95** | 11.979 | < 11.98 | ≈ neutral |
| p_oodc | 7.9 | 7.3 | **7.60** | 7.643 | < 7.65 | ≈ neutral |
| p_tan | 28.5 | 30.6 | **29.55** | 28.341 | < 28.34 | ❌ **+4.3% worse** |
| p_re | 6.4 | 6.4 | **6.40** | 6.300 | < 6.30 | ❌ **+1.6% worse** |

- **W&B runs:** h1lj2fej (seed 42), xw0xlhpc (seed 73)
- **W&B group:** blended-l1l2
- **Epochs completed:** 148 (s42), 147 (s73) — both hit 180-min wall-clock timeout
- **Peak GPU memory:** ~55 GiB (s42), ~72 GiB (s73)
- **val/loss:** 0.3798 (s42), 0.3783 (s73)

### Command used
```bash
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py --agent alphonse --wandb_name "alphonse/blended-l1l2-alpha01" \
  --wandb_group "blended-l1l2" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --l2_surface_weight 0.1 \
  --seed 42
```

### What happened

**The hypothesis did not work.** The blended L1+L2 loss hurt performance on the two most challenging splits:

- **p_tan (tandem transfer) is 4.3% worse.** This is the most significant regression. The tandem transfer split involves geometry configurations not seen in training, and the L2 penalty appears to over-penalize large errors during training, causing the model to become more conservative on these hard out-of-distribution cases. The high inter-seed variance on p_tan (28.5 vs 30.6) also suggests the L2 component makes training less stable.
- **p_re is 1.6% worse.** The OOD Reynolds number split also degraded, though less dramatically.
- **p_in and p_oodc are approximately neutral** — within noise of the baseline.

The core issue is likely that the L2 gradient scales linearly with error size, which in theory should help the worst-predicted nodes. In practice, however, this appears to create a competing gradient signal with the existing asymmetric hard-node mining (which already upweights high-error surface nodes at epoch 30+), leading to over-correction on tandem samples. The tandem_ramp + adaptive_boost already handles the tandem/non-tandem imbalance, and adding L2 on top may be double-penalizing the hardest cases.

Additionally, the L2 was applied across all 3 channels (Ux, Uy, p), while the L1 surface loss only operates on pressure (channel 2). This mismatch means the L2 gradient included velocity components that weren't part of the primary surface optimization signal, potentially diluting the pressure-focused learning.

### Suggested follow-ups
- If revisiting blended losses, apply L2 only to the pressure channel to match L1 scope
- Try a much smaller alpha (0.01-0.03) to reduce interference with existing hard-node mining
- Consider a scheduled alpha that ramps from 0 to target over training, avoiding early-training interference
- The existing hard-node mining + tandem_ramp mechanisms may already achieve what the L2 was intended to do — additional large-error penalties may be redundant

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2250 because ADVISOR: Closing — the blended L1+L2 loss didn't work. The 2-seed avg shows p_tan +4.3% worse (29.55 vs 28.34 baseline) and p_re +1.6% worse (6.40 vs 6.30). Your analysis is spot-on: the L2 gradient conflicts with the existing hard-node mining + tandem_ramp mechanisms, effectively double-penalizing the hardest cases. The channel mismatch (L2 on Ux/Uy/p vs L1 only on p) is also a valid concern.

Key takeaway: the baseline's existing asymmetric error-weighting mechanisms (PCGrad extreme + tandem ramp) are already quite effective at directing gradient to high-error nodes. Additional quadratic penalties are redundant and destabilizing.

Good observation about the high inter-seed variance on p_tan (28.5 vs 30.6) — another signal that the L2 is hurting generalization stability. Moving you to a fresh direction shortly..
